File: dsa/dsa.cpp
/**
 * Athena Library Management System - Core DSA C++ Implementations
 * This file provides standard, native C++ classes parallel to the JavaScript structures.
 * It contains a demonstration suite running test cases for verification.
 */

#include <iostream>
#include <vector>
#include <string>
#include <list>
#include <algorithm>
#include <stdexcept>
#include <memory>
#include <sstream>
#include <functional>

// ==========================================
// 1. DATA RECORD TYPES
// ==========================================

struct Book {
    int id;
    std::string title;
    std::string author;
    std::string isbn;
    int quantity;
    std::string category;
};

struct Transaction {
    std::string id;
    std::string timestamp;
    std::string type; // "issue", "return", "queue", "queue_handoff", "add_book", "register_member"
    int bookId;
    std::string bookTitle;
    std::string userId;
    std::string userName;
    std::string details;
};

struct QueueUser {
    std::string userId;
    std::string name;
    std::string dateAdded;
};


// ==========================================
// 2. STACK: LIFO History Audit Ledger
// ==========================================
class HistoryStack {
private:
    std::vector<Transaction> items;
    const size_t MAX_SIZE = 50;

public:
    HistoryStack() = default;

    // Push new audit transaction onto top of LIFO Stack
    void push(const Transaction& item) {
        if (items.size() >= MAX_SIZE) {
            items.erase(items.begin()); // Remove oldest element (shift boundary)
        }
        items.push_back(item);
    }

    // Pop top audit record (triggers rollback)
    Transaction pop() {
        if (isEmpty()) {
            throw std::underflow_error("History audit stack registers are empty.");
        }
        Transaction top = items.back();
        items.pop_back();
        return top;
    }

    Transaction peek() const {
        if (isEmpty()) {
            throw std::underflow_error("History audit stack is empty.");
        }
        return items.back();
    }

    bool isEmpty() const {
        return items.empty();
    }

    size_t size() const {
        return items.size();
    }

    std::vector<Transaction> toArray() const {
        return items;
    }

    void clear() {
        items.clear();
    }
};


// ==========================================
// 3. QUEUE: FIFO Waitlist System
// ==========================================
class WaitingQueue {
private:
    int bookId;
    std::vector<QueueUser> items;

public:
    explicit WaitingQueue(int id) : bookId(id) {}

    // Enqueue a student to the rear/tail of FIFO Waitlist
    bool enqueue(const QueueUser& user) {
        if (hasUser(user.userId)) return false; // Prevent duplicates
        items.push_back(user);
        return true;
    }

    // Dequeue student at the front/head of the queue for book dispatching
    QueueUser dequeue() {
        if (isEmpty()) {
            throw std::underflow_error("Waitlist queue is empty.");
        }
        QueueUser front = items.front();
        items.erase(items.begin());
        return front;
    }

    bool hasUser(const std::string& userId) const {
        for (const auto& item : items) {
            if (item.userId == userId) return true;
        }
        return false;
    }

    bool removeUser(const std::string& userId) {
        auto it = std::find_if(items.begin(), items.end(), 
            [&userId](const QueueUser& user) { return user.userId == userId; });
        if (it != items.end()) {
            items.erase(it);
            return true;
        }
        return false;
    }

    bool isEmpty() const {
        return items.empty();
    }

    size_t size() const {
        return items.size();
    }

    std::vector<QueueUser> toArray() const {
        return items;
    }
};


// ==========================================
// 4. TREE: Hierarchical Category Index Directory
// ==========================================
struct TreeNode {
    std::string name;
    std::string path;
    std::vector<std::shared_ptr<TreeNode>> children;

    TreeNode(std::string nm, std::string pth) : name(nm), path(pth) {}
};

class CategoryTree {
private:
    std::shared_ptr<TreeNode> root;

    // Recursive helper to scan for matching path
    std::shared_ptr<TreeNode> findNode(const std::string& path, std::shared_ptr<TreeNode> node) {
        if (node->path == path) return node;
        for (auto child : node->children) {
            auto found = findNode(path, child);
            if (found) return found;
        }
        return nullptr;
    }

    void collectPaths(std::shared_ptr<TreeNode> node, std::vector<std::string>& list) {
        if (node != root) {
            list.push_back(node->path);
        }
        for (auto child : node->children) {
            collectPaths(child, list);
        }
    }

public:
    CategoryTree() {
        root = std::make_shared<TreeNode>("Root", "");
    }

    // Traverse tree structure to insert subcategory leaf
    bool addCategory(const std::string& parentPath, const std::string& childName) {
        auto parentNode = findNode(parentPath, root);
        if (!parentNode) return false;

        std::string childPath = parentPath.empty() ? childName : parentPath + "/" + childName;
        for (auto child : parentNode->children) {
            if (child->name == childName) return false; // Collision check
        }

        parentNode->children.push_back(std::make_shared<TreeNode>(childName, childPath));
        return true;
    }

    // Purge category node and child subcategories recursively
    bool deleteCategory(const std::string& path) {
        if (path.empty()) return false;

        size_t idx = path.find_last_of('/');
        std::string nameToDelete = (idx == std::string::npos) ? path : path.substr(idx + 1);
        std::string parentPath = (idx == std::string::npos) ? "" : path.substr(0, idx);

        auto parentNode = findNode(parentPath, root);
        if (!parentNode) return false;

        auto& vec = parentNode->children;
        for (auto it = vec.begin(); it != vec.end(); ++it) {
            if ((*it)->name == nameToDelete) {
                vec.erase(it);
                return true;
            }
        }
        return false;
    }

    std::vector<std::string> getAllPaths() {
        std::vector<std::string> paths;
        collectPaths(root, paths);
        return paths;
    }
};


// ==========================================
// 5. HASH MAP: Prime-Modulus Buckets Search Index
// ==========================================
struct HashNode {
    std::string key;
    Book book;
};

class BookHashMap {
private:
    int size;
    std::vector<std::list<HashNode>> buckets;

    // Polynomial rolling hash function modulus prime size
    int hash(const std::string& key) {
        int hashVal = 0;
        const int prime = 31;
        for (char ch : key) {
            hashVal = (hashVal * prime + tolower(ch)) % size;
        }
        return hashVal;
    }

public:
    explicit BookHashMap(int sz = 13) : size(sz), buckets(sz) {}

    // Put item mapping into collision bucket list
    void put(const std::string& key, const Book& book) {
        std::string normKey = key;
        std::transform(normKey.begin(), normKey.end(), normKey.begin(), ::tolower);
        int index = hash(normKey);

        auto& bucket = buckets[index];
        for (const auto& node : bucket) {
            if (node.book.id == book.id && node.key == normKey) return;
        }
        bucket.push_back({normKey, book});
    }

    // Retrieve matching records
    std::vector<Book> get(const std::string& key) {
        std::string normKey = key;
        std::transform(normKey.begin(), normKey.end(), normKey.begin(), ::tolower);
        int index = hash(normKey);

        std::vector<Book> matches;
        for (const auto& node : buckets[index]) {
            if (node.key == normKey || node.key.find(normKey) != std::string::npos) {
                matches.push_back(node.book);
            }
        }
        return matches;
    }
};


// ==========================================
// 6. SORTING: Bubble & Quick Sort Algorithms
// ==========================================
class SortingAlgorithms {
public:
    // Bubble Sort: O(N^2) sorting
    template<typename T>
    static void bubbleSort(std::vector<T>& arr, std::function<bool(const T&, const T&)> compare) {
        int n = arr.size();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (compare(arr[j], arr[j + 1])) {
                    std::swap(arr[j], arr[j + 1]);
                }
            }
        }
    }

private:
    template<typename T>
    static int partition(std::vector<T>& arr, int low, int high, std::function<bool(const T&, const T&)> compare) {
        T pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (compare(arr[j], pivot)) {
                i++;
                std::swap(arr[i], arr[j]);
            }
        }
        std::swap(arr[i + 1], arr[high]);
        return i + 1;
    }

public:
    // Quick Sort: O(N log N) recursive sorting
    template<typename T>
    static void quickSort(std::vector<T>& arr, int low, int high, std::function<bool(const T&, const T&)> compare) {
        if (low < high) {
            int pi = partition(arr, low, high, compare);
            quickSort(arr, low, pi - 1, compare);
            quickSort(arr, pi + 1, high, compare);
        }
    }
};


// ==========================================
// 7. DEMONSTRATION RUNNER
// ==========================================
int main() {
    std::cout << "=========================================================\n";
    std::cout << " Athena DSA Library System - C++ Core Test Sandbox       \n";
    std::cout << "=========================================================\n\n";

    // 1. HistoryStack test
    std::cout << "[Test 1] history stack (LIFO)\n";
    HistoryStack history;
    history.push({ "tx_01", "10:00:00", "issue", 201, "Algorithms", "1001", "Rahul Sharma", "Issued Volume" });
    history.push({ "tx_02", "10:05:00", "issue", 202, "AI Modern Approach", "1002", "Aarav Mehta", "Issued Volume" });

    std::cout << "Stack size: " << history.size() << "\n";
    std::cout << "Top stack item: " << history.peek().bookTitle << " (" << history.peek().userName << ")\n";
    
    auto popped = history.pop();
    std::cout << "Popped stack item: " << popped.bookTitle << "\n";
    std::cout << "New stack size: " << history.size() << "\n\n";

    // 2. WaitingQueue test
    std::cout << "[Test 2] waiting queue (FIFO)\n";
    WaitingQueue queue(203); // Clean Code waitlist
    queue.enqueue({ "1002", "Aarav Mehta", "10:00" });
    queue.enqueue({ "1003", "Priya Patel", "10:02" });
    
    std::cout << "Queue size: " << queue.size() << "\n";
    auto front = queue.dequeue();
    std::cout << "Dequeued front user: " << front.name << "\n";
    std::cout << "New queue size: " << queue.size() << "\n\n";

    // 3. CategoryTree test
    std::cout << "[Test 3] hierarchical category tree\n";
    CategoryTree tree;
    tree.addCategory("", "Computer Science");
    tree.addCategory("Computer Science", "DSA");
    tree.addCategory("Computer Science", "AI");
    tree.addCategory("", "Mathematics");

    auto paths = tree.getAllPaths();
    std::cout << "Generated Category Tree Paths:\n";
    for (const auto& path : paths) {
        std::cout << "  - " << path << "\n";
    }
    std::cout << "\n";

    // 4. BookHashMap test
    std::cout << "[Test 4] priming hash map index search\n";
    BookHashMap hashIndex(13);
    Book book1 = { 201, "Introduction to Algorithms", "Thomas Cormen", "9780262033848", 2, "Computer Science/DSA" };
    Book book2 = { 202, "Artificial Intelligence", "Stuart Russell", "9780136042594", 1, "Computer Science/AI" };
    hashIndex.put("Algorithms", book1);
    hashIndex.put("AI", book2);
    hashIndex.put("Thomas Cormen", book1);

    auto matches = hashIndex.get("Algorithms");
    std::cout << "Search 'Algorithms' returned: " << matches.size() << " match(es).\n";
    if (!matches.empty()) {
        std::cout << "  Matched Title: " << matches[0].title << "\n";
    }
    std::cout << "\n";

    // 5. SortingAlgorithms test
    std::cout << "[Test 5] quick sorting algorithm\n";
    std::vector<Book> sortList = { book2, book1 }; // AI has ID 202, DSA has ID 201
    std::cout << "Before sorting:\n";
    for (const auto& b : sortList) std::cout << "  - ID " << b.id << ": " << b.title << "\n";

    auto compareByIdAsc = [](const Book& a, const Book& b) {
        return a.id < b.id;
    };
    
    SortingAlgorithms::quickSort<Book>(sortList, 0, sortList.size() - 1, compareByIdAsc);

    std::cout << "After Quick Sort (ID Ascending):\n";
    for (const auto& b : sortList) std::cout << "  - ID " << b.id << ": " << b.title << "\n";
    std::cout << "\n";

    std::cout << "All core DSA verification cases pass.\n";
    std::cout << "=========================================================\n";
    return 0;
}

File: dsa/dsa.js
/**
 * DSA Library Management System - Core Data Structures
 * This file contains custom implementations of core DSA concepts for both business logic and visual playback.
 */

// ==========================================
// 1. STACK: Borrow History Tracking
// ==========================================
class HistoryStack {
    constructor(initialItems = []) {
        this.items = [...initialItems];
    }

    push(item) {
        // Limit history to last 50 entries to keep memory clean
        if (this.items.length >= 50) {
            this.items.shift(); // Remove oldest
        }
        this.items.push({
            id: 'txn_' + Date.now() + '_' + Math.random().toString(36).substr(2, 5),
            timestamp: new Date().toLocaleString(),
            ...item
        });
    }

    pop() {
        return this.items.pop();
    }

    peek() {
        return this.items[this.items.length - 1];
    }

    isEmpty() {
        return this.items.length === 0;
    }

    toArray() {
        return [...this.items];
    }

    clear() {
        this.items = [];
    }
}

// ==========================================
// 2. QUEUE: Waiting List System (FIFO)
// ==========================================
class WaitingQueue {
    constructor(bookId, initialItems = []) {
        this.bookId = bookId;
        this.items = [...initialItems]; // Array of { userId, name, dateAdded }
    }

    enqueue(user) {
        // Prevent duplicate queue entries for the same user-book combination
        if (this.hasUser(user.userId)) return false;
        
        this.items.push({
            userId: user.userId,
            name: user.name,
            dateAdded: new Date().toLocaleString()
        });
        return true;
    }

    dequeue() {
        if (this.isEmpty()) return null;
        return this.items.shift();
    }

    peek() {
        if (this.isEmpty()) return null;
        return this.items[0];
    }

    isEmpty() {
        return this.items.length === 0;
    }

    size() {
        return this.items.length;
    }

    hasUser(userId) {
        return this.items.some(item => item.userId === userId);
    }

    removeUser(userId) {
        const index = this.items.findIndex(item => item.userId === userId);
        if (index !== -1) {
            return this.items.splice(index, 1)[0];
        }
        return null;
    }

    toArray() {
        return [...this.items];
    }
}

// ==========================================
// 3. TREE: Category Tree Structure
// ==========================================
class TreeNode {
    constructor(name, path) {
        this.name = name;
        this.path = path; // e.g. "Computer Science/DSA"
        this.children = [];
    }
}

class CategoryTree {
    constructor(savedTree = null) {
        if (savedTree) {
            this.root = this.deserialize(savedTree);
        } else {
            this.root = new TreeNode("Root", "");
            this.initializeDefaultCategories();
        }
    }

    initializeDefaultCategories() {
        this.addCategory("", "Computer Science");
        this.addCategory("Computer Science", "DSA");
        this.addCategory("Computer Science", "AI");
        this.addCategory("Computer Science", "Web Development");

        this.addCategory("", "Mathematics");
        this.addCategory("Mathematics", "Calculus");
        this.addCategory("Mathematics", "Algebra");

        this.addCategory("", "Physics");
        this.addCategory("Physics", "Quantum");
        this.addCategory("Physics", "Mechanics");
    }

    // Find node by path
    findNode(path, node = this.root) {
        if (!path) return this.root;
        if (node.path === path) return node;

        for (let child of node.children) {
            const found = this.findNode(path, child);
            if (found) return found;
        }
        return null;
    }

    // Add child to a parent path
    addCategory(parentPath, childName) {
        const parentNode = this.findNode(parentPath);
        if (!parentNode) return false;

        const childPath = parentPath ? `${parentPath}/${childName}` : childName;
        
        // Prevent duplicate child name
        if (parentNode.children.some(c => c.name === childName)) return false;

        const newNode = new TreeNode(childName, childPath);
        parentNode.children.push(newNode);
        return true;
    }

    // Delete node and its subcategories
    deleteCategory(path) {
        if (!path) return false;
        
        const parts = path.split('/');
        const nameToDelete = parts.pop();
        const parentPath = parts.join('/');
        
        const parentNode = this.findNode(parentPath);
        if (!parentNode) return false;

        const index = parentNode.children.findIndex(c => c.name === nameToDelete);
        if (index !== -1) {
            parentNode.children.splice(index, 1);
            return true;
        }
        return false;
    }

    // Helper to traverse and build list of categories
    getAllPaths(node = this.root, list = []) {
        if (node !== this.root) {
            list.push(node.path);
        }
        for (let child of node.children) {
            this.getAllPaths(child, list);
        }
        return list;
    }

    // Serialize tree structure to clean JSON object
    serialize(node = this.root) {
        return {
            name: node.name,
            path: node.path,
            children: node.children.map(child => this.serialize(child))
        };
    }

    // Deserialize tree from JSON object
    deserialize(data) {
        const node = new TreeNode(data.name, data.path);
        node.children = data.children.map(childData => this.deserialize(childData));
        return node;
    }
}

// ==========================================
// 4. HASH MAP: Fast Search System (ISBN/Title/Author)
// ==========================================
class BookHashMap {
    constructor(size = 13) { // Size 13 (prime) to demonstrate collisions beautifully
        this.size = size;
        this.buckets = Array(size).fill(null).map(() => []);
    }

    // Visualizable Hash Function (Polynomial rolling hash or simple string sum)
    hash(key) {
        let hashVal = 0;
        const prime = 31;
        for (let i = 0; i < Math.min(key.length, 10); i++) {
            hashVal = (hashVal * prime + key.charCodeAt(i)) % this.size;
        }
        return hashVal;
    }

    // Put a book in the hash map (can map a single key to multiple books - e.g. multi-value search index)
    put(key, book) {
        const normalizedKey = key.toLowerCase().trim();
        const index = this.hash(normalizedKey);
        
        const bucket = this.buckets[index];
        // Check if book already exists in bucket, to avoid duplicates
        const exists = bucket.some(item => item.book.id === book.id && item.key === normalizedKey);
        if (!exists) {
            bucket.push({ key: normalizedKey, book: book });
        }
    }

    // Rebuild hash map index from complete book list
    rebuildIndex(books) {
        this.buckets = Array(this.size).fill(null).map(() => []);
        
        books.forEach(book => {
            // Index by ID
            this.put(book.id.toString(), book);
            // Index by ISBN
            if (book.isbn) this.put(book.isbn, book);
            // Index by Title
            if (book.title) {
                this.put(book.title, book);
                // Also index by individual words in title for elastic-like search
                book.title.split(/\s+/).forEach(word => {
                    if (word.length > 2) this.put(word, book);
                });
            }
            // Index by Author
            if (book.author) {
                this.put(book.author, book);
                book.author.split(/\s+/).forEach(word => {
                    if (word.length > 2) this.put(word, book);
                });
            }
        });
    }

    // Get matching books for a search key, returning hashing lookup trace steps
    get(key) {
        const normalizedKey = key.toLowerCase().trim();
        const index = this.hash(normalizedKey);
        const bucket = this.buckets[index];
        
        const matches = [];
        const trace = {
            searchKey: key,
            hashIndex: index,
            bucketSize: bucket.length,
            comparisons: []
        };

        for (let i = 0; i < bucket.length; i++) {
            const item = bucket[i];
            trace.comparisons.push({
                slotIndex: i,
                comparedKey: item.key,
                match: item.key === normalizedKey || item.key.includes(normalizedKey) || normalizedKey.includes(item.key)
            });
            
            if (item.key === normalizedKey || item.key.includes(normalizedKey) || normalizedKey.includes(item.key)) {
                // Ensure no duplicate books are added to match list
                if (!matches.some(m => m.id === item.book.id)) {
                    matches.push(item.book);
                }
            }
        }

        return { matches, trace };
    }
}

// ==========================================
// 5. SORTING: Visual Merge/Quick Sort Tracker
// ==========================================
class SortingAlgorithms {
    // Generate comparisons & swaps for visual playback in UI
    
    // 5a. Visual Bubble Sort (Highly visual for step-by-step swaps)
    static async bubbleSort(arr, compareFn, onStep) {
        let items = [...arr];
        let n = items.length;
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n - i - 1; j++) {
                // Highlighting index j and j+1
                await onStep([...items], [j, j + 1], []);
                
                if (compareFn(items[j], items[j + 1]) > 0) {
                    // Swapping
                    let temp = items[j];
                    items[j] = items[j + 1];
                    items[j + 1] = temp;
                    
                    // Highlight swapped items
                    await onStep([...items], [], [j, j + 1]);
                }
            }
        }
        await onStep([...items], [], []);
        return items;
    }

    // 5b. Visual Quick Sort (Divide & Conquer)
    static async quickSort(arr, compareFn, onStep) {
        let items = [...arr];
        
        const partition = async (low, high) => {
            let pivot = items[high];
            let i = low - 1;
            
            for (let j = low; j < high; j++) {
                // j and high (pivot) are being compared
                await onStep([...items], [j, high], []);
                
                if (compareFn(items[j], pivot) < 0) {
                    i++;
                    let temp = items[i];
                    items[i] = items[j];
                    items[j] = temp;
                    await onStep([...items], [], [i, j]);
                }
            }
            
            let temp = items[i + 1];
            items[i + 1] = items[high];
            items[high] = temp;
            await onStep([...items], [], [i + 1, high]);
            return i + 1;
        };

        const quickSortHelper = async (low, high) => {
            if (low < high) {
                let pi = await partition(low, high);
                await quickSortHelper(low, pi - 1);
                await quickSortHelper(pi + 1, high);
            }
        };

        await quickSortHelper(0, items.length - 1);
        await onStep([...items], [], []);
        return items;
    }

    // Comparison functions for Books
    static getCompareFunction(sortBy, order = 'asc') {
        const multiplier = order === 'asc' ? 1 : -1;
        
        return (a, b) => {
            let valA, valB;
            if (sortBy === 'title') {
                valA = a.title.toLowerCase();
                valB = b.title.toLowerCase();
            } else if (sortBy === 'author') {
                valA = a.author.toLowerCase();
                valB = b.author.toLowerCase();
            } else if (sortBy === 'quantity') {
                valA = Number(a.quantity);
                valB = Number(b.quantity);
            } else {
                valA = a.id;
                valB = b.id;
            }

            if (valA < valB) return -1 * multiplier;
            if (valA > valB) return 1 * multiplier;
            return 0;
        };
    }
}
